tokio/src/io/unix/pipe.rs

```diff
@@ -0,0 +1,1191 @@
+    /// loop for reading.
+    ///
+    /// # Errors
+    ///
+    /// If the file type check fails, this function will fail with `io::ErrorKind::Other`.
+    /// This function may also fail with other standard OS errors.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn open_receiver<P: AsRef<Path>>(&self, path: P) -> io::Result<Receiver> {
+        let file = self.open(path.as_ref(), PipeEnd::Receiver)?;
+        Receiver::from_file_unchecked(file)
+    }
+
+    /// Creates a [`Sender`] from a FIFO file with the options specified by `self`.
+    ///
+    /// This function will open the FIFO file at the specified path, possibly
+    /// check if it is a pipe, and associate the pipe with the default event
+    /// loop for writing.
+    ///
+    /// # Errors
+    ///
+    /// If the file type check fails, this function will fail with `io::ErrorKind::Other`.
+    /// If the file is not opened in read-write access mode and the file is not
+    /// currently open for reading, this function will fail with `ENXIO`.
+    /// This function may also fail with other standard OS errors.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn open_sender<P: AsRef<Path>>(&self, path: P) -> io::Result<Sender> {
+        let file = self.open(path.as_ref(), PipeEnd::Sender)?;
+        Sender::from_file_unchecked(file)
+    }
+
+    fn open(&self, path: &Path, pipe_end: PipeEnd) -> io::Result<File> {
+        let mut options = std::fs::OpenOptions::new();
+        options
+            .read(pipe_end == PipeEnd::Receiver)
+            .write(pipe_end == PipeEnd::Sender)
+            .custom_flags(libc::O_NONBLOCK);
+
+        #[cfg(target_os = "linux")]
+        if self.read_write {
+            options.read(true).write(true);
+        }
+
+        let file = options.open(path)?;
+
+        if !self.unchecked && !is_fifo(&file)? {
+            return Err(io::Error::new(io::ErrorKind::Other, "not a pipe"));
+        }
+
+        Ok(file)
+    }
+}
+
+impl Default for OpenOptions {
+    fn default() -> OpenOptions {
+        OpenOptions::new()
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+enum PipeEnd {
+    Sender,
+    Receiver,
+}
+
+cfg_net_unix! {
+    /// Writing end of a Unix pipe.
+    ///
+    /// # Examples
+    ///
```

Comment:
This doesn't need to be wrapped in `cfg_net_unix!`. The entire file is already wrapped in that.

Reply:
That's true, but I noticed that most types in `net` are wrapped like this, even though, just like here, entire modules are wrapped in macros, for example see UnixStream. I thought that maybe it's a coding convention to throw a compile error if the type is not placed in a right place, since `cfg_net_unix` only wraps a type without its impl block. I'll remove it if that's not the case.